tower-hedge/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+#[macro_use]
+extern crate futures;
+extern crate hdrhistogram;
+#[macro_use]
+extern crate log;
+extern crate tokio_timer;
+extern crate tower_service;
+
+use futures::{Async, Future, Poll};
+use hdrhistogram::Histogram;
+use tokio_timer::{clock, Delay};
+use tower_service::Service;
+
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+mod rotating;
+
+use rotating::Rotating;
+
+/// A "retry policy" to classify if a request should be pre-emptively retried.
+pub trait Policy<Request>: Sized {
+    /// Check the policy if a certain request should be pre-emptively retried.
+    ///
+    /// This method is passed a reference to the original request.
+    fn can_retry(&self, req: &Request) -> bool;
+    /// Tries to clone a request before being passed to the inner service.
+    ///
+    /// If the request cannot be cloned, return `None`.
+    fn clone_request(&self, req: &Request) -> Option<Request>;
+}
+
+/// A middleware that pre-emptively retries requests which have been outstanding
+/// for longer than a given latency percentile.  If either of the original
+/// future or the retry future completes, that value is used.
+#[derive(Clone)]
+pub struct Hedge<P, S> {
```

Comment:
It would be cool if `Hedge` could be a layer that always fires off two requests and the `Policy` describes how to build the second request.

In the specific case that you are handling here, could the histogram / latency tracking be moved into the policy? Can the policy trait be:

```rust
pub trait Policy<Request> {
    type Response;
    type Error;
    type Future;

    fn hedge(&mut self, req: &Request) -> Self::Future;
}
```

This looks *very* much like `Service`, but it takes a `&Request` instead. I wonder if there could be a way to somehow "just" use `Service` as a policy, but that isn't initially important.

The way you may be able to implement hedge w/ the histogram would be to return a response future that waits for the delay before sending the request. It could be that there should be a `tower-delay` layer that does this. We can punt though.

Does structuring it like this seem possible (or a good idea)?

Reply:
Yes, I think this sounds possible.  However, you'd still need some middleware above or below to measure response latency and feed that data to the policy.  Is it better to assemble this out of small components (latency histogram layer, delay, percentile based policy, and the hedge itself) or have an all-in-one solution like I have here?